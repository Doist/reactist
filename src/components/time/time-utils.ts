import dayjs from 'dayjs'
/**
 * There's a problem with our setup where the default export from
 * localizedFormat (and likely every other dayjs plugin) isn't properly
 * recognized. The proposed workarounds (importing with `.js` ending, or adding
 * `allowSyntheticDefaultImports` to the tsconfig) either broke linting or type
 * checking. After spending some time on this it was decided that further
 * investigations are not worth it, the code works and the eslint ignore is fine.
 * ref: https://github.com/iamkun/dayjs/issues/593
 * ref: https://day.js.org/docs/en/installation/typescript
 */
// eslint-disable-next-line import/default
import LocalizedFormat from 'dayjs/plugin/localizedFormat'

dayjs.extend(LocalizedFormat)

type TimeConfig = {
    locale?: string
    longFormat?: string
    shortFormatCurrentYear?: string
    shortFormatPastYear?: string
    daysSuffix?: string
    hoursSuffix?: string
    minutesSuffix?: string
    momentsAgo?: string
}

const TimeUtils = {
    SHORT_FORMAT_CURRENT_YEAR: 'L',
    SHORT_FORMAT_PAST_YEAR: 'LL',
    LONG_FORMAT: 'LL, LT',

    timeAgo(timestamp: number, config: TimeConfig = {}) {
        const {
            locale = 'en',
            shortFormatCurrentYear = this.SHORT_FORMAT_CURRENT_YEAR,
            shortFormatPastYear = this.SHORT_FORMAT_PAST_YEAR,
            daysSuffix = 'd',
            hoursSuffix = 'h',
            minutesSuffix = 'm',
            momentsAgo = 'moments ago',
        } = config
        const now = dayjs()
        const date = dayjs(timestamp * 1000)
        date.locale(locale)
        const diffMinutes = now.diff(date, 'minute')
        const diffHours = now.diff(date, 'hour')
        const diffDays = now.diff(date, 'day')

        if (diffDays > 1) {
            if (date.isSame(now, 'year')) {
                return date.format(shortFormatCurrentYear)
            } else {
                return date.format(shortFormatPastYear)
            }
        } else if (diffDays === 1) {
            return `${diffDays}${daysSuffix}`
        } else if (diffHours > 0 && diffHours <= 23) {
            return `${diffHours}${hoursSuffix}`
        } else if (diffMinutes > 0 && diffMinutes <= 59) {
            return `${diffMinutes}${minutesSuffix}`
        } else {
            return momentsAgo
        }
    },

    formatTime(timestamp: number, config: TimeConfig = {}) {
        const {
            locale = 'en',
            shortFormatCurrentYear = this.SHORT_FORMAT_CURRENT_YEAR,
            shortFormatPastYear = this.SHORT_FORMAT_PAST_YEAR,
        } = config
        const date = dayjs(timestamp * 1000)
        date.locale(locale)
        if (date.isSame(dayjs(), 'year')) {
            return date.format(shortFormatCurrentYear)
        } else {
            return date.format(shortFormatPastYear)
        }
    },

    formatTimeLong(timestamp: number, config: TimeConfig = {}) {
        const { locale = 'en', longFormat = this.LONG_FORMAT } = config
        const date = dayjs(timestamp * 1000)
        date.locale(locale)
        return date.format(longFormat)
    },
}

export type { TimeConfig }
export { TimeUtils }
