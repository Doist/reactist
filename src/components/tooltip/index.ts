export { Tooltip, SHOW_DELAY, HIDE_DELAY } from './tooltip'
export type { TooltipProps } from './tooltip'
