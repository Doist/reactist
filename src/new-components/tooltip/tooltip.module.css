.tooltip {
    text-overflow: ellipsis;
    z-index: var(--reactist-stacking-order-tooltip);
}
