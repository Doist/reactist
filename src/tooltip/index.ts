export { Tooltip, TooltipProvider } from './tooltip'
export type { TooltipProps } from './tooltip'
