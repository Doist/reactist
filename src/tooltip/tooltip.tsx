import * as React from 'react'

import {
    useTooltipStore,
    Tooltip as AriakitTooltip,
    TooltipAnchor,
    TooltipArrow,
} from '@ariakit/react'
import { Box } from '../box'

import type { TooltipStoreState } from '@ariakit/react'

import styles from './tooltip.module.css'

type TooltipProps = {
    /**
     * The element that triggers the tooltip. Generally a button or link.
     *
     * It should be an interactive element accessible both via mouse and keyboard interactions.
     */
    children: React.ReactNode

    /**
     * The content to show in the tooltip.
     *
     * It can be rich content provided via React elements, or string content. It should not include
     * interactive elements inside it. This includes links or buttons.
     *
     * You can provide a function instead of the content itself. In this case, the function should
     * return the desired content. This is useful if the content is expensive to generate. It can
     * also be useful if the content dynamically changes often, so every time you trigger the
     * tooltip the content may have changed (e.g. if you show a ticking time clock in the tooltip).
     *
     * The trigger element will be associated to this content via `aria-describedby`. This means
     * that the tooltip content will be read by assistive technologies such as screen readers. It
     * will likely read this content right after reading the trigger element label.
     */
    content: React.ReactNode | (() => React.ReactNode)

    /**
     * How to place the tooltip relative to its trigger element.
     *
     * The possible values are "top", "bottom", "left", "right". Additionally, any of these values
     * can be combined with `-start` or `-end` for even more control. For instance `top-start` will
     * place the tooltip at the top, but with the start (e.g. left) side of the toolip and the
     * trigger aligned. If neither `-start` or `-end` are provided, the tooltip is centered along
     * the vertical or horizontal axis with the trigger.
     *
     * The position is enforced whenever possible, but tooltips can appear in different positions
     * if the specified one would make the tooltip intersect with the viewport edges.
     *
     * @default 'top'
     */
    position?: TooltipStoreState['placement']

    /**
     * The separation (in pixels) between the trigger element and the tooltip.
     * @default 3
     */
    gapSize?: number

    /**
     * Whether to show an arrow-like element attached to the tooltip, and pointing towards the
     * trigger element.
     * @default false
     */
    withArrow?: boolean

    /**
     * An escape hatch, in case you need to provide a custom class name to the tooltip.
     */
    exceptionallySetClassName?: string
}

function Tooltip({
    children,
    content,
    position = 'top',
    gapSize = 3,
    withArrow = false,
    exceptionallySetClassName,
}: TooltipProps) {
    const tooltip = useTooltipStore({ placement: position, showTimeout: 500, hideTimeout: 100 })
    const isOpen = tooltip.useState('open')

    const child = React.Children.only(
        children as React.FunctionComponentElement<JSX.IntrinsicElements['div']> | null,
    )

    if (!child) {
        return child
    }

    if (typeof child.ref === 'string') {
        throw new Error('Tooltip: String refs cannot be used as they cannot be forwarded')
    }

    return (
        <>
            <TooltipAnchor render={child} store={tooltip} ref={child.ref} />
            {isOpen && content ? (
                <Box
                    as={AriakitTooltip}
                    gutter={gapSize}
                    store={tooltip}
                    className={[styles.tooltip, exceptionallySetClassName]}
                    background="toast"
                    borderRadius="standard"
                    paddingX="small"
                    paddingY="xsmall"
                    maxWidth="medium"
                    width="fitContent"
                    overflow="hidden"
                    textAlign="center"
                >
                    {withArrow ? <TooltipArrow /> : null}
                    {typeof content === 'function' ? content() : content}
                </Box>
            ) : null}
        </>
    )
}

export type { TooltipProps }
export { Tooltip }
