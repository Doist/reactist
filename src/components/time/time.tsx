import { Component } from 'react'
import { Tooltip } from '../../tooltip'
import { TimeUtils, type TimeConfig } from './time-utils'

import './time.less'

const DELAY = 60000

type Props = {
    /** UNIX timestamp of the time to display. */
    time?: number
    /** Configuration for localization. */
    config?: TimeConfig
    /** Additional css class applied to the time element. */
    className?: string
    tooltipOnHover?: boolean
    /** Refresh the component every DELAY seconds. */
    refresh?: boolean
    /** If you don't want to use the default time format on the tooltip use this prop to supply a custom text */
    tooltip?: React.ReactNode
    /** When hovering over time it expands to short absolute version. */
    expandOnHover?: boolean
    /** When hovering over time it expands to the full absolute version. */
    expandFullyOnHover?: boolean
}

type State = {
    hovered: boolean
    mouseX?: number
    mouseY?: number
}

class Time extends Component<Props, State> {
    public static displayName: string
    public static defaultProps: Props

    constructor(props: Props) {
        super(props)
        this.refreshInterval = undefined

        this.state = {
            hovered: false,
            mouseX: undefined,
            mouseY: undefined,
        }
    }

    componentDidMount() {
        if (this.props.refresh) {
            this._refresh()
        }
    }

    componentDidUpdate(prevProps: Props) {
        if (!prevProps.refresh && this.props.refresh) {
            this._refresh()
        }

        if (prevProps.refresh && !this.props.refresh) {
            if (this.refreshInterval) {
                clearTimeout(this.refreshInterval)
            }
        }
    }

    componentWillUnmount() {
        if (this.refreshInterval) {
            clearTimeout(this.refreshInterval)
        }
    }

    refreshInterval?: ReturnType<typeof setTimeout>

    _setHovered(hovered: boolean, event: React.MouseEvent) {
        const { mouseX, mouseY } = this.state
        const { clientX, clientY } = event
        if (clientX !== mouseX || clientY !== mouseY) {
            // mouse has moved
            this.setState(() => ({
                hovered,
                mouseX: clientX,
                mouseY: clientY,
            }))
        }
    }

    _renderTime(config: Props['config']) {
        if (!this.props.time) {
            return
        }
        if (this.state.hovered) {
            if (this.props.expandFullyOnHover && !this.props.tooltipOnHover) {
                return TimeUtils.formatTimeLong(this.props.time, config)
            }
            if (this.props.expandOnHover && !this.props.tooltipOnHover) {
                return TimeUtils.formatTime(this.props.time, config)
            }
        }
        return TimeUtils.timeAgo(this.props.time, config)
    }

    _refresh() {
        this.refreshInterval = setInterval(() => {
            this.forceUpdate()
        }, DELAY)
    }

    render() {
        let className = 'reactist_time'
        if (this.props.className) {
            className = this.props.className
        }

        const timeComponent = this._renderTime(this.props.config)

        return (
            <time
                className={className}
                onMouseEnter={(event) => this._setHovered(true, event)}
                onMouseLeave={(event) => this._setHovered(false, event)}
            >
                {this.props.tooltipOnHover ? (
                    <Tooltip
                        content={
                            this.props.tooltip ||
                            (this.props.time &&
                                TimeUtils.formatTimeLong(this.props.time, this.props.config))
                        }
                    >
                        <span>{timeComponent}</span>
                    </Tooltip>
                ) : (
                    timeComponent
                )}
            </time>
        )
    }
}
Time.displayName = 'Time'

Time.defaultProps = {
    expandOnHover: false,
    expandFullyOnHover: false,
    tooltipOnHover: false,
    refresh: true,
    config: {
        locale: 'en',
        daysSuffix: 'd',
        hoursSuffix: 'h',
        minutesSuffix: 'm',
        momentsAgo: 'moments ago',
    },
}

export { Time }
