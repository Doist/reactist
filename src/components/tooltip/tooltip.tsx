import * as React from 'react'
import classNames from 'classnames'

import {
    useTooltipState as useAriakitTooltipState,
    TooltipStateProps as AriakitTooltipStateProps,
    Tooltip as AriakitTooltip,
    TooltipProps as AriakitTooltipProps,
    TooltipAnchor,
    TooltipAnchorProps,
} from 'ariakit/tooltip'
import type { PopoverState } from 'ariakit/popover'

import './tooltip.less'

type TooltipProps = Omit<AriakitTooltipProps, 'children' | 'state'> & {
    children: React.ReactNode
    content: React.ReactNode | (() => React.ReactNode)
    position?: PopoverState['placement']
    gapSize?: number
}

// These are exported to be used in the tests, they are not meant to be exported publicly
export const SHOW_DELAY = 500
export const HIDE_DELAY = 100

function useDelayedTooltipState(initialState: AriakitTooltipStateProps) {
    const tooltipState = useAriakitTooltipState(initialState)
    const delay = useDelay()
    return React.useMemo(
        () => ({
            ...tooltipState,
            show: delay(() => tooltipState.show(), SHOW_DELAY),
            hide: delay(() => tooltipState.hide(), HIDE_DELAY),
        }),
        [delay, tooltipState],
    )
}

function Tooltip({
    children,
    content,
    position = 'top',
    gapSize = 3,
    className,
    ...props
}: TooltipProps) {
    const state = useDelayedTooltipState({ placement: position, gutter: gapSize })

    const child = React.Children.only(
        children as React.FunctionComponentElement<JSX.IntrinsicElements['div']> | null,
    )

    if (!content || !child) {
        return child
    }

    if (typeof child.ref === 'string') {
        throw new Error('Tooltip: String refs cannot be used as they cannot be forwarded')
    }

    /**
     * Prevents the tooltip from automatically firing on focus all the time. This is to prevent
     * tooltips from showing when the trigger element is focused back after a popover or dialog that
     * it opened was closed. See link below for more details.
     * @see https://github.com/reakit/reakit/discussions/749
     */
    function handleFocus(event: React.FocusEvent<HTMLDivElement>) {
        // If focus is not followed by a key up event, does it mean that it's not
        // an intentional keyboard focus? Not sure but it seems to work.
        // This may be resolved soon in an upcoming version of reakit:
        // https://github.com/reakit/reakit/issues/750
        function handleKeyUp(e: Event) {
            const eventKey = (e as KeyboardEvent).key
            if (eventKey !== 'Escape' && eventKey !== 'Enter' && eventKey !== 'Space') {
                state.show()
            }
        }
        event.currentTarget.addEventListener('keyup', handleKeyUp, { once: true })
        // Prevent tooltip.show from being called by TooltipReference
        event.preventDefault()
        child?.props?.onFocus?.(event)
    }

    return (
        <>
            <TooltipAnchor state={state} ref={child.ref} onFocus={handleFocus}>
                {(anchorProps: TooltipAnchorProps) => {
                    // Let child props override anchor props so user can specify attributes like tabIndex
                    // Also, do not apply the child's props to TooltipAnchor as props like `as` can create problems
                    // by applying the replacement component/element twice
                    return React.cloneElement(child, { ...anchorProps, ...child.props })
                }}
            </TooltipAnchor>
            {state.visible ? (
                <AriakitTooltip
                    {...props}
                    state={state}
                    className={classNames('reactist_tooltip', className)}
                >
                    {typeof content === 'function' ? content() : content}
                </AriakitTooltip>
            ) : null}
        </>
    )
}

export type { TooltipProps }
export { Tooltip }

//
// Internal helpers
//

/**
 * Returns a function offering the same interface as setTimeout, but cleans up on unmount.
 *
 * The timeout state is shared, and only one delayed function can be active at any given time. If
 * a new delayed function is called while another one was waiting for its time to run, that older
 * invocation is cleared and it never runs.
 *
 * This is suitable for our use case here, but probably not the most intuitive thing in general.
 * That's why this is not made a shared util or something like it.
 */
function useDelay() {
    const timeoutRef = React.useRef<ReturnType<typeof setTimeout>>()

    const clearTimeouts = React.useCallback(function clearTimeouts() {
        if (timeoutRef.current != null) {
            clearTimeout(timeoutRef.current)
        }
    }, [])

    // Runs clearTimeouts when the component is unmounted
    React.useEffect(() => clearTimeouts, [clearTimeouts])

    return React.useCallback(
        function delay(fn: () => void, delay: number) {
            return () => {
                clearTimeouts()
                timeoutRef.current = setTimeout(fn, delay)
            }
        },
        [clearTimeouts],
    )
}
