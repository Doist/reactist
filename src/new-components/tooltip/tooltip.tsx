import * as React from 'react'
import classNames from 'classnames'
import { TooltipPopup, useTooltip } from '@reach/tooltip'
import styles from './tooltip.module.css'

type Position = 'top' | 'right' | 'bottom' | 'left'

type PositioningArgs = {
    triggerRect?: Partial<DOMRect> | null
    tooltipRect?: Partial<DOMRect> | null
    offset: number
    position: Position
}

/**
 * Override default positioning logic
 *
 * Original implementation: https://github.com/reach/reach-ui/blob/d4bcfc73f27e6e618ec2eb2b0154a58bc17496f0/packages/tooltip/src/index.tsx#L568
 * Centering demo: https://reacttraining.com/reach-ui/tooltip/#triangle-pointers-and-custom-styles
 */
function computePosition({
    triggerRect,
    tooltipRect,
    offset,
    position,
}: PositioningArgs): Pick<React.CSSProperties, 'top' | 'right' | 'bottom' | 'left'> {
    if (
        !triggerRect ||
        !tooltipRect ||
        typeof triggerRect.top !== 'number' ||
        typeof triggerRect.left !== 'number' ||
        typeof triggerRect.bottom !== 'number' ||
        typeof triggerRect.right !== 'number' ||
        typeof triggerRect.width !== 'number' ||
        typeof triggerRect.height !== 'number' ||
        typeof tooltipRect.width !== 'number' ||
        typeof tooltipRect.height !== 'number'
    ) {
        return {}
    }

    // const collisions = getCollisions(offset, triggerRect, tooltipRect)
    const collisions = {
        top: triggerRect.top - tooltipRect.height < 0,
        right: window.innerWidth < triggerRect.left + tooltipRect.width,
        bottom: window.innerHeight < triggerRect.bottom + tooltipRect.height + offset,
        left: triggerRect.left - tooltipRect.width < 0,
    }

    if (position === 'top' || position === 'bottom') {
        if (position === 'bottom' && collisions.bottom && !collisions.top) {
            position = 'top'
        } else if (position === 'top' && collisions.top && !collisions.bottom) {
            position = 'bottom'
        }

        const triggerCenter = triggerRect.left + triggerRect.width / 2
        const left = triggerCenter - tooltipRect.width / 2
        const maxLeft = window.innerWidth - tooltipRect.width - offset

        return {
            left: Math.min(Math.max(offset, left), maxLeft) + window.scrollX,
            top:
                position === 'top'
                    ? triggerRect.top - offset - window.scrollY - tooltipRect.height
                    : triggerRect.bottom + offset + window.scrollY,
        }
    } else {
        if (position === 'left' && collisions.left && !collisions.right) {
            position = 'right'
        } else if (position === 'right' && collisions.right && !collisions.left) {
            position = 'left'
        }

        const triggerCenter = triggerRect.top + triggerRect.height / 2
        const top = triggerCenter - tooltipRect.height / 2
        const maxTop = window.innerHeight - tooltipRect.height - offset

        return {
            top: Math.min(Math.max(offset, top), maxTop) + window.scrollX,
            left:
                position === 'left'
                    ? triggerRect.left - offset - window.scrollY - tooltipRect.width
                    : triggerRect.right + offset + window.scrollY,
        }
    }
}

/**
 * Used to prevent the tooltip from automatically firing on focus all the time. This is to prevent
 * tooltips from showing when the trigger element is focused back after a popover or dialog that
 * it opened was closed. See link below for more details.
 * @see https://github.com/reakit/reakit/discussions/749
 */
function useSpecialFocusManagement(
    onFocus: React.FocusEventHandler | null | undefined,
): [boolean, React.FocusEventHandler] {
    const [preventShow, setPreventShow] = React.useState(false)

    function handleFocus(event: React.FocusEvent<HTMLDivElement>) {
        // If focus is not followed by a key up event, does it mean that it's not
        // an intentional keyboard focus? Not sure but it seems to work.
        function handleKeyUp(e: Event) {
            const eventKey = (e as KeyboardEvent).key
            if (eventKey !== 'Escape' && eventKey !== 'Enter' && eventKey !== 'Space') {
                setPreventShow(false)
            }
        }
        event.currentTarget.addEventListener('keyup', handleKeyUp, { once: true })
        setPreventShow(true)
        onFocus?.(event)
    }

    return [preventShow, handleFocus]
}

type ChildProps = Pick<
    JSX.IntrinsicElements['div'],
    | 'onMouseEnter'
    | 'onMouseMove'
    | 'onMouseLeave'
    | 'onFocus'
    | 'onBlur'
    | 'onKeyDown'
    | 'onMouseDown'
>

type TooltipProps = {
    id?: string
    children: React.ReactNode
    content: React.ReactNode | (() => React.ReactNode)
    position?: Position
    offset?: number
    exceptionallySetClassName?: string
}

/**
 * @see https://github.com/reach/reach-ui/blob/d4bcfc73f27e6e618ec2eb2b0154a58bc17496f0/packages/tooltip/src/index.tsx#L363
 */
function Tooltip({
    id,
    children,
    position = 'top',
    offset = 3,
    content,
    exceptionallySetClassName,
}: TooltipProps) {
    const child = React.Children.only<React.ReactElement<ChildProps>>(
        children as React.ReactElement<ChildProps>,
    )
    const [preventShow, handleFocus] = useSpecialFocusManagement(child.props.onFocus)

    const [trigger, tooltip] = useTooltip({
        id,
        onMouseEnter: child.props.onMouseEnter,
        onMouseMove: child.props.onMouseMove,
        onMouseLeave: child.props.onMouseLeave,
        onFocus: handleFocus,
        onBlur: child.props.onBlur,
        onKeyDown: child.props.onKeyDown,
        onMouseDown: child.props.onMouseDown,
        /* eslint-disable */
        ref: (child as any).ref,
        /* eslint-enable */
        // DEBUG_STYLE,
    })

    const positioningFunction = React.useCallback(
        function positioningFunction(
            triggerRect?: Partial<DOMRect> | null,
            tooltipRect?: Partial<DOMRect> | null,
        ) {
            return computePosition({ triggerRect, tooltipRect, offset, position })
        },
        [offset, position],
    )

    const label =
        tooltip.isVisible && !preventShow
            ? content instanceof Function
                ? content()
                : content
            : null

    return (
        <>
            {React.cloneElement(children as React.ReactElement, trigger)}
            {label ? (
                <TooltipPopup
                    className={classNames(styles.tooltip, exceptionallySetClassName)}
                    position={positioningFunction}
                    label={label}
                    {...tooltip}
                />
            ) : null}
        </>
    )
}

export { Tooltip }
export type { TooltipProps }
