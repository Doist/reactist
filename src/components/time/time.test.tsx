import dayjs from 'dayjs'
import * as React from 'react'
import { render, screen, waitFor } from '@testing-library/react'
import { axe } from 'jest-axe'

import { Time } from './time'
import userEvent from '@testing-library/user-event'

import { SHOW_DELAY } from '../../tooltip/tooltip'

describe('Time', () => {
    beforeAll(() => {
        jest.useFakeTimers()
        jest.setSystemTime(new Date('2023-03-14T12:00:00.000Z'))
    })

    afterAll(() => {
        jest.useRealTimers()
    })

    const testDate = dayjs(new Date('March 22, 1991 13:37:42')).unix()

    it('renders without crashing', () => {
        const { container } = render(<Time time={dayjs().unix()} />)
        expect(container).toMatchSnapshot()
    })

    it('toggles hovered state on mouse enter and leave when mouse moves', () => {
        render(<Time time={testDate} expandFullyOnHover />)

        expect(screen.getByText('March 22, 1991')).toBeVisible()

        userEvent.hover(screen.getByText('March 22, 1991'))
        expect(screen.getByText('March 22, 1991, 1:37 PM')).toBeVisible()

        // <Time> checks that the mouse coordinates have changed before setting state
        userEvent.unhover(screen.getByText('March 22, 1991, 1:37 PM'), { clientX: 10, clientY: 10 })
        expect(screen.getByText('March 22, 1991')).toBeVisible()
    })

    it('does not toggle hovered state when mouse did not move', () => {
        render(<Time time={testDate} expandFullyOnHover />)

        expect(screen.getByText('March 22, 1991')).toBeVisible()

        userEvent.hover(screen.getByText('March 22, 1991'))
        expect(screen.getByText('March 22, 1991, 1:37 PM')).toBeVisible()

        // <Time> checks that the mouse coordinates have changed before setting state
        userEvent.unhover(screen.getByText('March 22, 1991, 1:37 PM'))
        expect(screen.getByText('March 22, 1991, 1:37 PM')).toBeVisible()
    })

    it('renders relative time when not hovered', () => {
        render(<Time time={dayjs().unix()} />)
        expect(screen.getByText('moments ago')).toBeVisible()
    })

    it('renders short absolute time when hovered and expandedOnHover is set', () => {
        render(<Time time={testDate} expandOnHover />)

        userEvent.hover(screen.getByText('March 22, 1991'))
        expect(screen.getByText('March 22, 1991')).toBeVisible()
    })

    it('adds additional class name if supplied', () => {
        const { container } = render(<Time time={testDate} className="this-classes were-added" />)
        expect(container).toMatchSnapshot()
    })

    it('renders wrapped in tooltip when tooltipOnHover is set', async () => {
        render(<Time time={testDate} tooltipOnHover />)

        userEvent.hover(screen.getByText('March 22, 1991'))
        await waitFor(
            () => {
                expect(
                    screen.getByRole('tooltip', { name: 'March 22, 1991, 1:37 PM' }),
                ).toBeVisible()
            },
            { timeout: SHOW_DELAY + 10 },
        )

        userEvent.unhover(screen.getByText('March 22, 1991'))
        await waitFor(() => {
            expect(screen.queryByRole('tooltip')).not.toBeInTheDocument()
        })
    })

    it('renders with custom tooltip when supplied', async () => {
        render(<Time time={testDate} tooltipOnHover tooltip="Test" />)

        userEvent.hover(screen.getByText('March 22, 1991'))
        await waitFor(
            () => {
                expect(screen.getByRole('tooltip', { name: 'Test' })).toBeVisible()
            },
            { timeout: SHOW_DELAY + 10 },
        )

        userEvent.unhover(screen.getByText('March 22, 1991'))
        await waitFor(() => {
            expect(screen.queryByRole('tooltip')).not.toBeInTheDocument()
        })
    })

    it('does not render short absolute time on hover when tooltipOnHover is set', async () => {
        render(<Time time={dayjs().unix()} tooltipOnHover expandOnHover />)
        userEvent.hover(screen.getByText('moments ago'))

        await waitFor(
            () => {
                expect(
                    screen.getByRole('tooltip', { name: dayjs().format('LL, LT') }),
                ).toBeVisible()
                expect(screen.getByText('moments ago')).toBeVisible()
            },
            { timeout: SHOW_DELAY + 10 },
        )

        userEvent.unhover(screen.getByText('moments ago'))
        await waitFor(() => {
            expect(screen.queryByRole('tooltip')).not.toBeInTheDocument()
        })
    })

    it('does not render full absolute time on hover when tooltipOnHover is set', async () => {
        render(<Time time={dayjs().unix()} tooltipOnHover expandFullyOnHover />)
        userEvent.hover(screen.getByText('moments ago'))

        await waitFor(
            () => {
                expect(
                    screen.getByRole('tooltip', { name: dayjs().format('LL, LT') }),
                ).toBeVisible()
                expect(screen.getByText('moments ago')).toBeVisible()
            },
            { timeout: SHOW_DELAY + 10 },
        )

        userEvent.unhover(screen.getByText('moments ago'))
        await waitFor(() => {
            expect(screen.queryByRole('tooltip')).not.toBeInTheDocument()
        })
    })

    describe('a11y', () => {
        beforeAll(() => {
            jest.useRealTimers()
        })

        it('renders with no a11y violations', async () => {
            const { container } = render(<Time time={dayjs().unix()} />)
            const results = await axe(container)

            expect(results).toHaveNoViolations()
        })
    })
})
