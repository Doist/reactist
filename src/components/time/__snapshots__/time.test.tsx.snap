// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Time adds additional class name if supplied 1`] = `
<div>
  <time
    class="this-classes were-added"
  >
    March 22, 1991
  </time>
</div>
`;

exports[`Time renders without crashing 1`] = `
<div>
  <time
    class="reactist_time"
  >
    moments ago
  </time>
</div>
`;
